### DISCUSSION TITLE
About ciclic references

### DISCUSSION THREAD

**Discussion by francipvb** [microsoft/pyright#7583] [Q&A]
Hello all,

I want to ask you about cyclic type references related to self-referencing and forward references to union types.

For example, this works at definition time, but doesn't when the type definition is used in a call:

```python
from typing import Union, Dict, List


Scalar = Union[str, float, int, bool]
Object = Dict[str, "JSONValue"]
Array = List["JSONValue"]
JSONValue = Union[Scalar, Object, Array, None]
```

And this pattern doesn't even work at runtime because an obvious reason:

```python
from __future__ import annotations

from typing import Union, Dict, List


Scalar = Union[str, float, int, bool]
Object = Dict[str, JSONValue]
Array = List[JSONValue]
JSONValue = Union[Scalar, Object, Array, None]
```

This is OK at definition but when you want to use the types you get an error. For example:

```python
from x.types import JSONValue


def parse_json(raw: str) -> JSONValue:
    ...


parse_json("null")  # pyright gives an error here
```

I can address this error for my own code, but third-party libs use this pattern to do it's typing support.

My question is: are you aware of this issue? Is there a tracking issue for this?

I am aware of 3.12 changes to this issue but the libraries cann't adapt to these changes without dropping support for python 3.11 and less.

Thank you,

**Comment by erictraut** [comment]
As you noted, the new `type` keyword in 3.12 eliminates this issue.

```python
type Scalar = str | float | int | bool
type Object = dict[str, "JSONValue"]
type Array = list["JSONValue"]
type JSONValue = Scalar | Object | Array | None
```

The `type` keyword isn't available in versions prior to 3.12, but you can manually allocate a `TypeAliasType` (which is what the `type` statement does under the covers). The `TypeAliasType` class is available through the `typing_extensions` backward compatibility library.

```python
from typing_extensions import TypeAliasType

JSONValue = TypeAliasType("JSONValue", "Scalar | Object | Array | None")
Scalar = TypeAliasType("Scalar", str | float | int | bool)
Object = TypeAliasType("Object", dict[str, "JSONValue"])
Array = TypeAliasType("Array", list["JSONValue"])
```

Another approach is to use the (now deprecated) `TypeAlias` annotation introduced in [PEP 613](https://peps.python.org/pep-0613/). It allows the type alias to be quoted so it is not evaluated by the runtime.

```python
from typing import TypeAlias

Scalar: TypeAlias = str | float | int | bool
Object: TypeAlias = dict[str, "JSONValue"]
Array: TypeAlias = list["JSONValue"]
JSONValue: TypeAlias = Scalar | Object | Array | None
```
